### PR TITLE
Deprecate `ember-mocha`

### PR DESCRIPTION
**Pull request by bertdeblock**
[Rendered RFC](https://github.com/bertdeblock/rfcs/blob/deprecate-ember-mocha/text/0858-deprecate-ember-mocha.md)

### CODE REVIEW

**Comment by jrjohnson** [comment]
This biggest motivation for me in this change would be removing a confusing branching decision when creating a new ember project. Should I use Mocha? Is it better? What are the pros and cons vs qunit? Are all questions caused by the presence of both of these official testing libs. Removing ember-mocha saves developers from a wheel spinning annoyance very early in their ember experience.

File: text/0858-deprecate-ember-mocha.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+References to `ember-mocha` should be removed from all learning materials, for example:
+
+- https://guides.emberjs.com/release/testing/testing-tools/#toc_mocha-chai-dom
+- https://guides.emberjs.com/release/testing/testing-tools/#toc_summary
+- https://guides.emberjs.com/release/testing/#toc_how-to-filter-tests
+
+## Drawbacks
+
+People using `ember-mocha` will have to migrate to using `ember-qunit` instead
```
Comment (wagenet) [inline text/0858-deprecate-ember-mocha.md:69]: They'll have to migrate anyway unless someone starts maintaining it, and it doesn't sound like people are leaping at that chance.

**Review by wagenet — APPROVED** [review]
Seems straightforward to me. We should discuss it at this week's meeting.

**Comment by ijlee2** [comment]
I'm in favor of deprecating `ember-mocha` for a few reasons:

- We help different Ember projects converge to a standard. When there is a problem with testing, solutions may be easier to share when there is common ground.
- We help simplify Ember.js code, as described in the proposal.
- We don't necessitate work on addon authors for a feature parity. For example, `ember-exam` has [a few features that are not supported in Mocha](https://github.com/ember-cli/ember-exam/tree/v8.0.0).

I imagine, how easily one can migrate from Mocha to QUnit partly depends on if the test suite had been well-maintained (i.e. the experience will be different from one team to another). The [`mocha-to-qunit`](https://github.com/freshworks/ember-freshdesk-codemods/tree/master/transforms/mocha-to-qunit) codemod might help with the migration.

**Comment by joegaudet** [comment]
Will there be code mods or other tools built to help those that have built code on ember-mocha? I believe Tom assured us this ember was a big tent when it came to testing?

**Comment by joegaudet** [comment]
(Also I prefer BDD style :P)

**Comment by bertdeblock** [comment]
The [mocha-to-qunit](https://github.com/freshworks/ember-freshdesk-codemods/tree/master/transforms/mocha-to-qunit) codemod was already suggested [here](https://github.com/emberjs/rfcs/pull/858#issuecomment-1308514390), so maybe that's an option?

**Comment by ijlee2** [comment]
@joegaudet @bertdeblock 

I do think it's unfortunate that there aren't many learning resources, to help teams see how to incrementally migrate a (large) test suite.

I found these links today:

- https://medium.com/@shibulijack/how-we-achieved-green-builds-in-an-enterprise-ember-js-app-be55431ac23e
- https://github.com/ember-codemods/ember-mocha-codemods

The Medium article mentions an in-repo addon that had allowed the team to run Mocha and QUnit side-by-side, but I wasn't able to find the addon if it had been open-sourced. @bertdeblock, maybe you could reach out to the author?

In theory, one would first run `ember-mocha-codemods` to migrate to recommended testing patterns in Ember v3, then `mocha-to-qunit`. Everything just works.

In practice, a more effective strategy may be to turn off Mocha completely:

1. If tests are run in continuous integration, temporarily set the test script in `package.json` to `exit 0` (unconditional success).

1. Uninstall `ember-mocha` and related packages. Install instead `ember-qunit`, `eslint-plugin-qunit`, `qunit`, and `qunit-dom` (refer to the Ember blueprint for [`package.json`](https://github.com/ember-cli/ember-cli/blob/v4.9.0-beta.0/blueprints/app/files/package.json)). Update [`.eslintrc.js`](https://github.com/ember-cli/ember-cli/blob/v4.9.0-beta.0/blueprints/app/files/.eslintrc.js#L63-L67) and [`tests/test-helper.js`](https://github.com/ember-cli/ember-cli/blob/v4.9.0-beta.0/blueprints/app/files/tests/test-helper.ts#L3-L6). For a sanity check, visit [localhost:4200/tests](http://localhost:4200/tests).

1. Introduce _default_ tests in `tests/integration` and `tests/units`, for every file that is important, e.g. components, routes, etc. Default tests are ones that Ember CLI creates when you run `ember g component-test ...`, `ember g route-test ...`, etc. They often include a tautology such as `assert.ok(true);`. The default tests guarantee that files can be read in tests and running their code doesn't cause an error.

1. Update the test script to run tests in continuous integration.

1. Re-introduce the tests that had been written in Mocha.

Each step corresponds to a separate pull request. Steps 3 and 5 could be split into multiple tasks so that a team of people can work on them in parallel, without running into merge conflicts.

**Comment by joegaudet** [comment]
It'd be much nicer if someone were able to adopt/update it, we run a rails
backend with all BDD style tests so that consistency is really helpful for
people looking to contribute in the front and back.

Obviously this is a big lift for someone, I'm no expert in testing
internals so supporting it seems out of my pay grade :P

.joe

**Comment by wagenet** [comment]
@joegaudet I don't think anyone is opposed to ember-mocha in concept, but the fact of the matter is that it's unmaintained and no one is stepping up to maintain it. At this point, it's better for us to say it's deprecated instead of acting like it's a viable option for people getting started.

I'm strongly opposed to just moving it into adopted addons unless we at least have someone who will make it compatible with Ember 4. To do so now would imply that it has support that it just doesn't have.

I don't love this outcome, but I'd prefer to at least be honest about where things stand.

**Comment by wagenet** [comment]
@joegaudet also, don't be so sure to assume the work is out of your pay grade. A lot of this stuff just ends up being a matter of getting your hands dirty and pushing through. If you do want and the only thing stopping you is feeling unqualified, let's see if we can help get you moving. Everyone starts somewhere!

**Comment by joegaudet** [comment]
Fair enough, I was mostly being cheeky, I've been here since the SproutCore
days :P

**Comment by joegaudet** [comment]
Is there a sense of what will be needed to bring it up to date?

.joe

**Comment by wagenet** [comment]
@joegaudet I should have realized the name was familiar! It sounds like ember-mocha doesn't even support Ember 4 yet, so that would be a good start.

File: text/0858-deprecate-ember-mocha.md (hunk 1 above)
Comment (bertdeblock) [reply text/0858-deprecate-ember-mocha.md:69]: This is great, thanks for sharing @octo-topi!

**Comment by octo-topi** [comment]
> The [mocha-to-qunit](https://github.com/freshworks/ember-freshdesk-codemods/tree/master/transforms/mocha-to-qunit?rgh-link-date=2022-11-18T08%3A23%3A41Z) codemod was already suggested [here](https://github.com/emberjs/rfcs/pull/858#issuecomment-1308514390), so maybe that's an option?

This codemod does the job https://github.com/1024pix/ember-freshdesk-codemods
We used in on a +/10kloc PR https://github.com/1024pix/pix/pull/5258

**Comment by octo-topi** [comment]
> The [mocha-to-qunit](https://github.com/freshworks/ember-freshdesk-codemods/tree/master/transforms/mocha-to-qunit?rgh-link-date=2022-11-18T08%3A23%3A41Z) codemod was already suggested [here](https://github.com/emberjs/rfcs/pull/858#issuecomment-1308514390), so maybe that's an option?

This codemod does the job [1024pix/ember-freshdesk-codemods](https://github.com/1024pix/ember-freshdesk-codemods)
We used in on a +/10kloc PR https://github.com/1024pix/pix/pull/5258

File: text/0858-deprecate-ember-mocha.md (hunk 1 above)
Comment (ijlee2) [reply text/0858-deprecate-ember-mocha.md:69]: @octo-topi I was amazed by the quality of your pull request. Texts can be auto-translated, no worries.

I think the one-to-one mapping between assertions in `mocha` and `qunit`, shown in [commit 3](https://github.com/1024pix/pix/pull/5258/commits/b0eccbdb63caee853d67d2b368c9f6079a334a08), will help other teams with their migration. 💯

Comment (octo-topi) [reply text/0858-deprecate-ember-mocha.md:69]: Well, @VincentHardouin is actually the one who has the job done, helped by several other members.
As we have an open-source repository, it was worth share the result so other people can focus on less tedious task.

**Comment by yads** [comment]
FYI, for everyone who is looking to migrate away from ember-mocha, here's a codemod to update your test files. I've already used this to migrate a project I maintain, it was fairly painless.
https://github.com/yads/ember-mocha-to-qunit-codemod

**Comment by apellerano-pw** [comment]
I had tried a couple years back on my personal acct to [bump the mocha major version in ember-mocha](https://github.com/emberjs/ember-mocha/pull/216), but it reached a point where a breaking change felt inevitable.

I think this is the situation causing ember-mocha's disrepair:

1. It has to navigate a very long mocha upgrade path,
    - which itself has breaking changes;
    - at least one of which directly violates ember-mocha's current architecture.
2. You have to understand the separation of responsibilities across ember-cli, ember-test-helpers, and ember-mocha.
3. You have to understand the broccoli->embroider transition.
4. Your project had to consciously shun qunit at some point which has default option momentum in both ember-cli and ember documentation.

Anyone who steps up to lead ember-mocha is going to break someone else's tests. They also need a really solid understanding of Ember ecosystem plumbing. And lastly they had to opt-in to mocha in the first place (to have some skin in the game.) I think this basically brings the set of possible maintainers to near-zero.

Could anyone spell out what ember-mocha maintainership would even look like? Like what is the role of ember-mocha in the big picture, and how small of a lib should it be to accomplish its goals? Assuming you had to build it from scratch today (which is basically what we need) what would you be signing up to build?

**Comment by wagenet** [comment]
I think we want to do everything in this RFC except actually deprecate the library. We should, however, move it out of the main Ember org and make it clear that it does not support Ember v4 at the moment. If we can establish a new maintainer there's no reason that it can't continue to be used going forwards.

**Comment by bertdeblock** [comment]
Will try to update the wording soon.

**Comment by wagenet** [comment]
@bertdeblock it's seeming like nobody actively wants to take over. We should move forward assuming that there won't be anyone.

**Comment by bertdeblock** [comment]
Updated the RFC a bit, based on some of the comments on this PR.

**Comment by kategengler** [comment]
The framework team gave this a final review today and we've decided to put it in final comment period for the Accepted stage.

**Pull request merged**